### DISCUSSION TITLE
resolve function

### DISCUSSION THREAD

**Discussion by jdidion** [openwdl/wdl#426] [Requests for Comment]
One thing that has tripped up many people is concatenation of `File` and `String`. Specifically, it is not clear whether the following expressions should result in `File` or `String`:

* `File` + `String`
* `String` + `File`

I propose that we disallow all concatenations between `File` and `String`, and instead introduce a resolve function:

`File resolve(String|File, String|File)`

`resolve` accepts two arguments that are both assumed to be paths. The first argument may be a relative or absolute path; the second argument must be a relative path. `resolve` appends the two paths and returns the result as a `File`. Importantly, `resolve` ensures the two paths are joined by the system-specific path separator (i.e. `/` on *nix and `\` on Windows).

**Comment by patmagee** [comment]
Just seeing this now. I think this is a great idea and definitely is inline with how files are treated in every other language. The implicit suggestion in wdl that "a file is just a special string" leads to soooo many issues and inconsistent behaviour.  I would definitely +1 this proposal

**Comment by rhpvorderman** [comment]
Looks great! The way you describe it it brings Python's `os.path.join` to WDL. Which is very neat.

I would propose that this function can allow an arbitrary number of arguments, or an array. `File resolve(String|File, String|File,  ...)` or `File resolve(Array[String|File])`. That would allow you to: `File my_file = resolve('somewhere', 'over', 'the', 'rainbow')` which yields `somewhere/over/the/rainbow` on *nix. Or if that is easier to implement for the execution engines `File my_file = resolve(['somewhere', 'over', 'the', 'rainbow'])`

My only gripe is with the name. `resolving` sounds like some 'thinking' is involved (i.e. ability to check if path exists, understanding `..` etc.). I would prefer a name like `pathjoin`, which implies dumb, system-specific path separator aware behavior. Unless we of course want to understand `.` and `..`. 
`pathjoin` also has my preference because the relation to filesystem paths is immediately clear, also to less programming oriented people.

**Reply by geoffjentry** [reply]
re what @rhpvorderman said, I was initially going to point out that this is effectively what `java.nio.Path.resolve()` does and thus we might be in a semantic debate based on people's biases from their daily driver programming language but that's not the case. In the JVM world this is semantically more like `Paths.get()` if I'm understanding correctly. 

**NB** I don't think this should be called `get` 😄

**Comment by mr-c** [comment]
> ensures the two paths are joined by the system-specific path separator (i.e. `/` on *nix and `\` on Windows).

If OpenWDL has standardized on BASH, then implicitly you've standardized on POSIX, therefore the path separator should always be `/`, yes?
